USAGE.adoc

```diff
@@ -0,0 +1,255 @@
+
+.`java.lang.Boolean.parseBoolean(String)` Invocation Example
+[source,java]
+----
+void invokeParseBoolean(MethodCreator method) {
+  // String val = "true";
+  ResultHandle str = method.load("true");
+  MethodDescriptor parseBoolean = MethodDescriptor.ofMethod(Boolean.class, "parseBoolean", boolean.class, String.class);
+  // boolean result = Boolean.parseBoolean(val);
+  ResultHandle result = method.invokeStaticMethod(parseBoolean, str);
+  // System.out.printl(val)
+  Gizmo.systemOutPrintln(method, result);
+}
+----
+
+`BytecodeCreator#invokeInterfaceMethod()` is used to invoke a method of an interface.
+
+.`java.util.Collection#size()` Invocation Example
+[source,java]
+----
+void invokeSize(MethodCreator method, ResultHandle someCollection) {
+  MethodDescriptor size = MethodDescriptor.ofMethod(Collection.class, "size", int.class);
+  // System.out.printl(someCollection.size())
+  Gizmo.systemOutPrintln(method, method.invokeInterfaceMethod(size, someCollection));
+}
+----
+
+`BytecodeCreator#invokeVirtualMethod()` is used to invoke a _virtual_ method, i.e. a public, package-private and protected methods of a class.
+
+.`java.lan.String#toLowerCase()` Invocation Example
+[source,java]
+----
+void invokeToLowerCase(MethodCreator method) {
+  // String val = "HELLO";
+  ResultHandle str = method.load("HELLO");
+  MethodDescriptor toLowerCase = MethodDescriptor.ofMethod(String.class, "toLowerCase", String.class);
+  // String result = str.toLowerCase();
+  ResultHandle result = method.invokeVirtualMethod(toLowerCase, str);
+  // System.out.printl(result)
+  Gizmo.systemOutPrintln(method, result);
+}
+----
+
+`BytecodeCreator#invokeSpecialMethod()` is used to invoke private instance methods, superclass methods or constructors.
+
+.Private Method Invocation Example
+[source,java]
+----
+void invokeSuperToString(MethodCreator method) {
+  MethodDescriptor myPrivateMethod = MethodDescriptor.of(Foo.class,"privateMethod", String.class);
+  // String result = privateMethod();
+  ResultHandle result = method.invokeSpecialMethod(myPrivateMethod, method.getThis()); <1>
+  // System.out.printl(result)
+  Gizmo.systemOutPrintln(method, result);
+}
+----
+<1> `BytecodeCreator.getThis()` represents the current object
+
+=== Loops
+
+Sometimes you need to generate the bytecode to iterate over a collection of elements.
+There are two constructs that could be useful: `ForEachLoop` and `WhileLoop`.
+In the following snippet we're going to generate a bytecode to iterate over all elements of an `java.lang.Iterable` instance.
+
+.`ForEachLoop` Example
+[source,java]
+----
+void iterate(MethodCreator method, ResultHandle iterable) {
+    // for (Object element : iterable) { 
+    //   System.out.println(element);
+    // }
+    ForEachLoop loop = method.forEach(list);
+    BytecodeCreator block = loop.block();
+    Gizmo.systemOutPrintln(block, loop.element());
+}
+----
+
```

Comment:
Maybe add a note here that `ForEachLoop` only works for `Iterable`s and not for arrays?